### PR TITLE
VIRTS-1153 Cleaned up module installation code

### PR DESCRIPTION
**Pull request by uruwhy**
Modules can also be installed as entire packages and not just single files (see `app/extensions/execute/shellcode/shellcode.py`)

### CODE REVIEW

File: app/utility/base_extension.py [hunk 1]
```diff
@@ -16,3 +19,40 @@ def check_go_dependencies(self):
...
+            if not os.path.exists(dest_package_path):
+                os.makedirs(dest_package_path)
+
+            # Check if entire package is to be copied
+            if file == '*':
+                for dir_item in os.listdir(src_package_path):
+                    src_path = os.path.join(src_package_path, dir_item)
+                    if os.path.isfile(src_path):
+                        copyfile(src=src_path,
+                                 dst=os.path.join(dest_package_path, dir_item))
```
Comment (khyberspache) [inline app/utility/base_extension.py:37]: separate function

File: app/utility/base_extension.py [hunk 2]
```diff
@@ -16,3 +19,40 @@ def check_go_dependencies(self):
...
+
+    def remove_module_files(self, base_dir):
+        for file, pkg in self.files:
+            package_path = os.path.join(base_dir, 'gocat', pkg)
+            # Check if entire package is to be deleted
+            if file == '*':
+                for dir_item in os.listdir(package_path):
+                    full_path = os.path.join(package_path, dir_item)
+                    if os.path.isfile(full_path) and dir_item != 'load.go':
+                        os.remove(full_path)
```
Comment (khyberspache) [inline app/utility/base_extension.py:50]: separate function

File: app/utility/base_extension.py [hunk 3]
```diff
@@ -16,3 +19,40 @@ def check_go_dependencies(self):
...
+            else:
+                file_path = os.path.join(package_path, file)
+                if os.path.exists(file_path):
+                    os.remove(file_path)
+
+    def install_dependencies(self):
+        # Attempt to install all dependencies if they aren't fulfilled
+        pass
```
Comment (khyberspache) [inline app/utility/base_extension.py:58]: remove for now

File: app/sand_svc.py [hunk 4]
```diff
@@ -186,3 +142,17 @@ def _load_extension_module(self, root, file):
...
             self.log.error('Error loading extension=%s, %s' % (module, e))
+
+    async def _attempt_module_copy(self, name):
+        """Attempts to copy the module files. Returns True upon success, False otherwise."""
+
+        module = self.sandcat_extensions.get(name)
+        if module:
+            try:
+                module.copy_module_files(base_dir=self.sandcat_dir)
+                return True
```
Comment (khyberspache) [inline app/sand_svc.py:153]: might make more sense for the copy_module_files function to return whether it succeeded or failed

File: app/sand_svc.py [hunk 5]
```diff
@@ -56,10 +59,24 @@ async def dynamically_compile_library(self, headers):
...
+            files = [f for f in files if not f[0] == '.' and not f[0] == "_"]
+            dirs[:] = [d for d in dirs if not d[0] == '.' and not d[0] == "_"]
+            for file in files:
+                module = await self._load_extension_module(root, file)
+                if module:
+                    if module.check_go_dependencies():
+                        self.sandcat_extensions[file.split('.')[0]] = module
+                    else:
+                        module.install_dependencies()
+
```
[the remark this answers is not in the document]
Comment (uruwhy) [reply app/sand_svc.py:79]: How about 
```python
            for file in files:
                module = await self._load_extension_module(root, file)
                if module:
                    if module.check_go_dependencies() or module.install_dependencies():
                        self.sandcat_extensions[file.split('.')[0]] = module
                    else:
                        self.log.error('Failed to fulfill dependencies for module %s' % module)
```

and `install_dependencies()` will return False by default?

[the remark this answers is not in the document]
Comment (khyberspache) [reply app/sand_svc.py:79]: nice!

**Pull request merged**